render/vulkan/shaders/texture.frag

```diff
@@ -0,0 +1,25 @@
+#version 450
+
+layout(set = 0, binding = 0) uniform sampler2D tex;
+
+layout(location = 0) in vec2 uv;
+layout(location = 0) out vec4 out_color;
+
+layout(push_constant) uniform UBO {
+	layout(offset = 80) float alpha;
+} data;
+
+void main() {
```

Comment:
Really, for such simple shaders, I think you should use compute shaders. They're mandatory, they're slightly faster than old OpenGL shaders due to not needing 2 programs and vertices, they're really so much simpler to work with, easily extendable to support whatever blending modes you need.

Reply:
I don't think that makes sense here. We support rendering with an arbitrary matrix and therefore need rasterization (think of a rotation transform for instance), emulating that via a compute shader isn't trivial, making it faster for all hardware would require quite some effort (if even possible at all). Vertex and fragment shaders are not "old OpenGL shaders".

Reply:
> We support rendering with an arbitrary matrix and therefore need rasterization (think of a rotation transform for instance), emulating that via a compute shader isn't trivial,

Technically, the code is no more complex. You are just doing the same calculation as in the vertex shader, but in the compute shader instead. The input to your vertex shader is basically a constant, so you can move the vertices to the compute shader if you really want to. Emulating it, even in the general case, *is* trivial enough that you can automate it with [a few lines of definitions](https://code.videolan.org/videolan/libplacebo/-/blob/master/src/dispatch.c#L969). I do this in my own code and it speeds up full-screen quads significantly compared to using the normal vertex/fragment infrastructure.

This may seem surprising, but compute shaders are closer to the hardware and for something like a full-screen quad where you have relatively simple vertices and full control over the pipeline, simply doing the calculation in the shader directly allows the implementation to optimize more than requiring it emulate those same computations when translating the vertex shader to something the GPU's compute can execute.

Though in this specific case it probably doesn't matter much either way.